File: src/java/magma/app/compile/value/Operation.java
package magma.app.compile.value;

import magma.api.option.None;
import magma.api.option.Option;
import magma.api.option.Some;
import magma.app.ValueCompiler;
import magma.app.compile.CompileState;
import magma.app.compile.type.PrimitiveType;
import magma.app.compile.type.Type;

public record Operation(Value left, String targetInfix, Value right) implements Value {
    @Override
    public String generate() {
        return ValueCompiler.getString(this.left) + " " + this.targetInfix + " " + ValueCompiler.getString(this.right);
    }

    public Option<Value> toValue() {
        return new Some<Value>(this);
    }

    public Type resolve(CompileState state) {
        return PrimitiveType.Unknown;
    }

    @Override
    public Option<String> generateAsEnumValue(String structureName) {
        return new None<String>();
    }
}

File: src/ts/magma/app/compile/value/Operation.ts
import { Node } from "../../../../magma/app/compile/node/Node";
import { ValueCompiler } from "../../../../magma/app/ValueCompiler";
import { Option } from "../../../../magma/api/option/Option";
import { Some } from "../../../../magma/api/option/Some";
import { CompileState } from "../../../../magma/app/compile/CompileState";
import { Primitives } from "../../../../magma/app/compile/type/Primitives";
export class Operation implements Node {
	left: Node;
	targetInfix: string;
	right: Node;
	constructor (left: Node, targetInfix: string, right: Node) {
		this.left = left;
		this.targetInfix = targetInfix;
		this.right = right;
	}
	generate(): string {
		return ValueCompiler.generateValue(this.left) + " " + this.targetInfix + " " + ValueCompiler.generateValue(this.right)/*unknown*/;
	}
	toNode(): Option<Node> {
		return new Some<Node>(this)/*unknown*/;
	}
	resolve(state: CompileState): Node {
		return Primitives.UNKNOWN/*unknown*/;
	}
	is(type: string): boolean {
		return false/*unknown*/;
	}
}
